lms/djangoapps/dashboard/sysadmin_get_course_users.py

```diff
@@ -0,0 +1,290 @@
+    """
+    if course_ordered[unit]['block_type'] in ['video', 'problem']:
+        if unit in visited_in_course:
+            status = '+'
+            return status
+        else:
+            status = '-'
+            return status
+    else:
+        current_seq_children_list = _get_children(sequential, course_ordered)
+        vert_position_in_seq = current_seq_children_list.index(vertical)
+        try:
+            last_visit_in_seq_position = seq_positions[sequential]
+            if (vert_position_in_seq < last_visit_in_seq_position):
+                status = '+'
+                return status
+            else:
+                status = '-'
+                return status
+        except Exception:
+            status = '-'
+            return status
+
+
+def get_report_data_for_course_users(courses, course_id):
+    """
+    Collects data for csv report with list of course users.
+
+    Args:
+        course_id: request.POST.get('course_id', '').strip()
+        courses: iterable list of courses from modulestore (xmodule.modulestore.django)
+
+    Returns:
+        course_users: dictionary {'header':header, 'data':data}
+    """
+    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    course = _find_course(course_key, courses)
+
+    if course is None:
+        message = 'Error - cannot get course with ID {}'.format(course_key)
+        return message
+    else:
+        data = []
+        enrolled_students = User.objects.filter(
+            courseenrollment__course_id=course_key)
+
+        # returns a list of chapters (ids)
+        # for parsing children (sequentials) in course tree
+        course_chapters = get_course_chapters(course_key)
+
+        course_obj = CourseStructure.objects.get(course_id=course_key)  # course content tree, serialised
+
+        # deserialising and converting to OrderedDict with the blocks
+        # in the order that they appear in the course
+        course_ordered = course_obj.ordered_blocks  # OrderedDict
+
+        # getting list of all sequentials (child for chapter) -
+        # The subsections defined in the course outline:
+        sequentials = _get_children_from_list(course_chapters, course_ordered)
+
+        # getting list of all verticals (child for sequential) -
+        # The units defined in the course outline:
+        verticals = _get_children_from_list(sequentials, course_ordered)
+
+        # getting vertical (unit) names
+        unit_names = [course_ordered[v].get('display_name') for v in verticals]
+
+        header = [_('username'), _('email'), _('registration date'), _('enrolled courses'), ('last login'), 'last visit', ]
+
+        # preparing list for counting visits for course units (last raw in the table)
+        visit_count = ['', '', '', '', '', 'Visits:', ]
+
+        for name in unit_names:
+            header.append(name)  # adding unit names to the table header
+            visit_count.append(0)  # and initiation of status counter
```

Comment:
I think you can extract the logic of adding the unit names to the header into separate function